haskell/providers.bzl

```diff
@@ -218,3 +79,203 @@ GhcPluginInfo = provider(
+        else:
+            libs_by_filename[lib.basename] = lib_to_link
+        filenames.append(lib.basename)
+
+    # Deduplicate the library names. Make sure to preserve static linking order.
+    filenames = depset(
+        transitive = [depset(direct = [filename]) for filename in filenames],
+        order = "topological",
+    ).to_list()
+
+    return [
+        libs_by_filename[filename]
+        for filename in filenames
+    ]
+
+def get_ghci_extra_libs(hs, cc_info, dynamic = True, path_prefix = None):
+    """Get libraries appropriate for GHCi's linker.
+
+    GHC expects dynamic and static versions of the same library to have the
+    same library name. Static libraries for which this is not the case will be
+    symlinked to a matching name.
+
+    Furthermore, dynamic libraries will be symbolically linked into a common
+    directory to allow for less RPATH entries and to fix file extensions that
+    GHCi does not support.
+
+    GHCi can load PIC static libraries (-fPIC -fexternal-dynamic-refs) and
+    dynamic libraries. Preferring static libraries can be useful to reduce the
+    risk of exceeding the MACH-O header size limit on MacOS, and to reduce
+    build times by avoiding to generate dynamic libraries. However, this
+    requires GHCi to run with the statically linked rts library.
+
+    Args:
+      hs: Haskell context.
+      cc_info: Combined CcInfo provider of dependencies.
+      dynamic: (optional) Whether to prefer dynamic libraries.
+      path_prefix: (optional) Prefix for the entries in the generated library path.
+
+    Returns:
+      (libs, ghc_env):
+        libs: depset of File, the libraries that should be passed to GHCi.
+        ghc_env: dict, environment variables to set for GHCi.
+
+    """
+    fixed_lib_dir = target_unique_name(hs, "_ghci_libs")
+    libs_to_link = _get_unique_lib_files(cc_info)
+    libs = []
+    for lib_to_link in libs_to_link:
+        dynamic_lib = None
+        if lib_to_link.dynamic_library:
+            dynamic_lib = lib_to_link.dynamic_library
+        elif lib_to_link.interface_library:
+            # XXX: Do these work with GHCi?
+            dynamic_lib = lib_to_link.interface_library
+        static_lib = None
+        if lib_to_link.pic_static_library:
+            static_lib = lib_to_link.pic_static_library
+        elif lib_to_link.static_library and hs.toolchain.is_windows:
+            # NOTE: GHCi cannot load non-PIC static libraries, except on Windows.
+            static_lib = lib_to_link.static_library
+
+        if dynamic_lib:
+            dynamic_lib = symlink_dynamic_library(hs, dynamic_lib, fixed_lib_dir)
+        static_lib = mangle_static_library(hs, dynamic_lib, static_lib, fixed_lib_dir)
+
+        lib = static_lib if static_lib else dynamic_lib
+        if dynamic and dynamic_lib:
+            lib = dynamic_lib
+
+        if lib:
+            libs.append(lib)
+
+    # NOTE: We can avoid constructing these in the future by instead generating
```

Comment:
Let's turn this into a ticket and link to the ticket in the comment. It's a fairly important note, because some corner cases essentially require us to resolve it.

Reply:
Done #926